rbd: add support to create clone in different pools

Yuggupta27 opened the pull request:
Add support to restore a pvc into a different pool.

Assuming there are 2 pools namely "replicapool" and "testpool"
Workflow:
1. Create a PVC in a replicapool.
2. For cloning, a snapshot gets created in replicapool.
3. PVC restore operation can be done in a new pool named testpool.

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>

Updates: #1838

Yuggupta27 committed:
rbd: pass parent volume to undoSnapshotCloning function

as we are supporting the creation of clone to a new
pool we need to pass the correct parent volume
to cleanup the snapshot on parent volume.

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: Yug <yuggupta27@gmail.com>

Yuggupta27 committed:
rbd: retrieve parent pool name of child image

when clones are created in different pool,we
need to retrieve the parent pool to get the
information of the parent image.

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: Yug <yuggupta27@gmail.com>

Yuggupta27 committed:
rbd: open new ioctx connection

if the parent and child clones are in
different namespaces we need to open a new
ioctx for pools.

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: Yug <yuggupta27@gmail.com>

Yuggupta27 committed:
rbd: add support to create clone in different pool

added support to create image in different pool.
if the snapshot/rbd image exists in one pool we
can create a clone the clone of the rbd image to
a different pool.

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: Yug <yuggupta27@gmail.com>

Yuggupta27 committed:
rbd: verify if pool name is not empty

Validate Snapshot request to check if the
passed pool name is not empty.

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: Yug <yuggupta27@gmail.com>

Yuggupta27 committed:
rbd: Update pool name in image chain

While traversing image chain, the parent
image can be present in a different pool
that the one child is in. So, updating
pool name in the next itteration to
that of the Parent.

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: Yug <yuggupta27@gmail.com>

Yuggupta27 committed:
e2e: update listRBDImages to list images in different pool

made pool as a argument of listRBDImages to support
listing of rbd images in different pools.

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: Yug <yuggupta27@gmail.com>

Yuggupta27 committed:
e2e: add function to create rbd pool

added a new function to create rbd pool
in e2e.

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: Yug <yuggupta27@gmail.com>

Yuggupta27 committed:
e2e: add an option to set rbd storageclass name

added an option to set storageclass name.

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: Yug <yuggupta27@gmail.com>

Yuggupta27 committed:
e2e: move yaml filepath declaration to global

declared yaml filepath in gobal for reusing.

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: Yug <yuggupta27@gmail.com>

Yuggupta27 committed:
e2e: add helper function for clone validation

added a helper function to test clone creation
in a different pool.

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: Yug <yuggupta27@gmail.com>

Yuggupta27 committed:
e2e: add e2e to test clone in different pool

added a e2e to test clones in different
pool.

Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: Yug <yuggupta27@gmail.com>

Yuggupta27 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.19

Yuggupta27 commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.19

rook deployment failed with
 ```
X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute
```

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1089,7 +1097,7 @@ func (rv *rbdVolume) cloneRbdImageFromSnapshot(ctx context.Context, pSnapOpts *r
 
 	util.DebugLog(ctx, logMsg,
-		pSnapOpts, image, rv.imageFeatureSet.Names(), rv.Monitors)
+		pSnapOpts.String(), rv, rv.imageFeatureSet.Names(), rv.Monitors)
```
there should not be a need to call `.String()`, calls to this function should be automatic with the formatting `%s` or `%q`. Other calls to `.String()` are removed where the object is a pointer to rbdVolume or rbdSnapshot.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1103,12 +1111,18 @@ func (rv *rbdVolume) cloneRbdImageFromSnapshot(ctx context.Context, pSnapOpts *r
 	}
+	// Destroy only the ioctx, connection is needed for other operations.
+	defer func() {
+		defer rv.ioctx.Destroy()
```
no need to do this, it will automatically be done when `rv` is destroyed.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1044,7 +1044,7 @@ func (rv *rbdVolume) hasSnapshotFeature() bool {
 
 func (rv *rbdVolume) createSnapshot(ctx context.Context, pOpts *rbdSnapshot) error {
-	util.DebugLog(ctx, "rbd: snap create %s using mon %s", pOpts, pOpts.Monitors)
+	util.DebugLog(ctx, "rbd: snap create %s@%s using mon %s", rv, pOpts.String(), pOpts.Monitors)
```
`%s@%s` and passing `rv, pOpts.String()` **should** be the same as just `%s` and `pOpts`. If it is not, that would be a bug (or very large inconsistency) in the caller of `createSnapshot`.

Or, if this is really how a rbdSnapshot does not match with its actual rbd-pool representation, introduce `rbdSnapshot.getSpec()` that returns the `<parentPool>/<parentName>@snapname` format.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1056,7 +1056,7 @@ func (rv *rbdVolume) createSnapshot(ctx context.Context, pOpts *rbdSnapshot) err
 
 func (rv *rbdVolume) deleteSnapshot(ctx context.Context, pOpts *rbdSnapshot) error {
-	util.DebugLog(ctx, "rbd: snap rm %s using mon %s", pOpts, pOpts.Monitors)
+	util.DebugLog(ctx, "rbd: snap rm %s@%s using mon %s", rv, pOpts.String(), pOpts.Monitors)
```
see comment for previous commit

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -532,23 +532,26 @@ func deleteImage(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) er
+
 		err = vol.getImageInfo()
+		// FIXME: create and destroy the vol inside the loop.
+		// see https://github.com/ceph/ceph-csi/pull/1838#discussion_r598530807
```
This really is dangerous. Maybe file a cleanup PR to do this and rebase this PR on top of that.

Or, create an issue for this problem as this really needs correcting urgently.

Yuggupta27 replied:
Created an issue for the same https://github.com/ceph/ceph-csi/issues/2144 . Since this issue exists in the current implementation and not something that this PR introduces, IMO we don't need to block this PR for that and we can work on the cleanup after this PR is merged to avoid last-minute breakages and delay. What do you think @nixpanic ?

humblec replied:
yeah, imo, it can be a followup PR.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -566,6 +566,7 @@ func (rv *rbdVolume) getCloneDepth(ctx context.Context) (uint, error) {
 			depth++
 		}
 		vol.RbdImageName = vol.ParentName
+		vol.Pool = vol.ParentPool
```
Same issue as before, re-using a rbdVolume with different attributes is quite wrong. Needs cleanup urgently.

nixpanic commented on e2e/rbd_helper.go:
```diff
@@ -232,6 +234,162 @@ func kmsIsVault(kms string) bool {
 	return kms == "vault"
 }
 
+// nolint:gocyclo // reduce complexity
```
Please use helper function from the start. Our e2e code should have the same quality as the other components.

nixpanic commented on e2e/rbd_helper.go:
```diff
@@ -232,6 +234,162 @@ func kmsIsVault(kms string) bool {
+	}
+	wg.Wait()
+
+	failed := 0
```
use something like the following to reduce the complexity:

```go
if failed := logErrors(wgErrs); failed != 0 {
    return fmt.Errorf("creating snapshots failed, %d errors were logged", failed)
}

Yuggupta27 replied:
Thanks for the suggestion! Added that commit!

nixpanic commented on e2e/rbd.go:
```diff
@@ -1212,6 +1212,50 @@ var _ = Describe("RBD", func() {
 
+			By("create rbd clones in different pool", func() {
+				// snapshot beta is only supported from v1.17+
+				if k8sVersionGreaterEquals(f.ClientSet, 1, 17) {
```
Check the inverse, and use `e2e.Skip()` to continue with the next test. That makes it easier to drop checks like this for old/unsupported Kubernetes versions.

Yuggupta27 replied:
Sounds good.

Yuggupta27 committed:
e2e: add logErr function to log errors

logErr function logs all the ocured errors
with a message that is passed for occurence
of each error.

Co-authored-by: Niels de Vos <ndevos@redhat.com>
Co-authored-by: Madhu Rajanna <madhupr007@gmail.com>
Signed-off-by: Yug <yuggupta27@gmail.com>

Yuggupta27 commented:
CI failure due to #2146

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -114,8 +114,10 @@ type rbdVolume struct {
 	Topology            map[string]string
 	// DataPool is where the data for images in `Pool` are stored, this is used as the `--data-pool`
 	// 	 argument when the pool is created, and is not used anywhere else
```
@Yuggupta27 not introduced with this patch, but please feel free to correct the `DataPool` explanation intendation.

Yuggupta27 replied:
Sure, I'll send a seperate PR to fix the same

Yuggupta27 replied:
Here https://github.com/ceph/ceph-csi/pull/2154

humblec commented on e2e/rbd_helper.go:
```diff
@@ -30,12 +32,15 @@ func rbdOptions(pool string) string {
 		return nil
 	}
+	if name != "" {
+		sc.Name = name
```
@Yuggupta27 also keep `sc.Name` in proper spacing compared to outer condition check

Yuggupta27 commented:
/retest all

humblec commented:
@Yuggupta27 As discussed please get the comments addressed in other PR . will get this in for now! Thanks :+1:

Madhu-1 commented:
/retest all

Yuggupta27 commented:
/retest ci/centos/mini-e2e/k8s-1.19

Yuggupta27 commented:
> /retest ci/centos/mini-e2e/k8s-1.19

rook deployment failure

Yuggupta27 commented:
/retest all

humblec commented:
@Mergifyio rebase

Yuggupta27 commented:
/retest ci/centos/upgrade-tests-cephfs

Yuggupta27 commented:
/retest ci/centos/mini-e2e/k8s-1.19

Yuggupta27 commented:
/retest ci/centos/mini-e2e/k8s-1.19
